src/main/java/org/opentripplanner/pricing/transit/ticket/TransitTicket.java

```diff
@@ -0,0 +1,286 @@
+        this.maxFares = validForFares;
+        this.maxDistance = validForDistance;
+        this.availableFrom = availableFrom;
+        this.availableTo = availableTo;
+    }
+
+    public static TransitTicketBuilder builder(int id, BigDecimal standardPrice) {
+        return new TransitTicketBuilder(id, standardPrice);
+    }
+
+    public static final class TransitTicketBuilder {
+
+        private final int id;
+        private final BigDecimal standardPrice;
+        private int validForMinutes = NO_LIMIT;
+        private int validForFares = NO_LIMIT;
+        private int validForDistance = NO_LIMIT;
+        private LocalDateTime availableFrom = null;
+        private LocalDateTime availableTo = null;
+
+        private TransitTicketBuilder(int id, BigDecimal standardPrice) {
+            this.id = id;
+            this.standardPrice = standardPrice;
+        }
+
+        public TransitTicket build() {
+            return new TransitTicket(this.id, this.standardPrice, this.validForMinutes, this.validForFares,
+                    this.validForDistance, this.availableFrom, this.availableTo);
+        }
+
+        public TransitTicketBuilder setTimeLimit(int validForMinutes) {
+            this.validForMinutes = getCorrectedLimitValue(validForMinutes);
+            return this;
+        }
+
+        public TransitTicketBuilder setFaresNumberLimit(int validForFares) {
+            this.validForFares = getCorrectedLimitValue(validForFares);
+            return this;
+        }
+
+        public TransitTicketBuilder setDistanceLimit(int validForDistance) {
+            this.validForDistance = getCorrectedLimitValue(validForDistance);
+            return this;
+        }
+
+        private int getCorrectedLimitValue(int evaluatedLimit) {
+            return evaluatedLimit > 0 ? evaluatedLimit : NO_LIMIT;
+        }
+
+        public TransitTicketBuilder setAvailableFrom(LocalDateTime availableFrom) {
+            this.availableFrom = availableFrom;
+            return this;
+        }
+
+        public TransitTicketBuilder setAvailableTo(LocalDateTime availableTo) {
+            this.availableTo = availableTo;
+            return this;
+        }
+
+    }
+
+    public void addAllowedAgency(String agencyId) {
+        routePatterns.put(agencyId, new RoutePattern());
+        stopPatterns.put(agencyId, new StopPattern());
+    }
+
+    public RoutePattern getRoutePattern(String agencyId) {
+        return routePatterns.get(agencyId);
+    }
+
+    public StopPattern getStopPattern(String agencyId) {
+        return stopPatterns.get(agencyId);
+    }
+
+    public boolean isAvailable(LocalDateTime currentTimestamp) {
+        return (Objects.isNull(this.availableTo) || this.availableTo.isAfter(currentTimestamp)) &&
+                (Objects.isNull(this.availableFrom) || !this.availableFrom.isAfter(currentTimestamp));
+    }
+
+    public int getTotalMinutesWhenValid(int finishesAtMinute, List<TransitTripStage> tripStages) {
```

Comment:
Maybe it is worth creating a reversed copy of this list `tripStages` and using simpler `for each` loops inside `get...ComplaintTime` methods